sdks/go/pkg/beam/core/runtime/exec/translate.go

```diff
@@ -372,32 +386,43 @@ func (b *builder) makeLink(from string, id linkID) (Node, error) {
 
 			switch op {
 			case graph.ParDo:
-				n := &ParDo{UID: b.idgen.New(), Inbound: in, Out: out}
-				n.Fn, err = graph.AsDoFn(fn, graph.MainUnknown)
+				dofn, err := graph.AsDoFn(fn, graph.MainUnknown)
 				if err != nil {
 					return nil, err
 				}
-				n.PID = transform.GetUniqueName()
-
-				input := unmarshalKeyedValues(transform.GetInputs())
-				for i := 1; i < len(input); i++ {
-					// TODO(herohde) 8/8/2018: handle different windows, view_fn and window_mapping_fn.
-					// For now, assume we don't need any information in the pardo payload.
-
-					ec, wc, err := b.makeCoderForPCollection(input[i])
-					if err != nil {
-						return nil, err
+				switch urn {
+				case urnPairWithRestriction:
+					u = &PairWithRestriction{UID: b.idgen.New(), Fn: dofn, Out: out[0]}
+				case urnSplitAndSizeRestrictions:
+					u = &SplitAndSizeRestrictions{UID: b.idgen.New(), Fn: dofn, Out: out[0]}
+				default:
+					n := &ParDo{UID: b.idgen.New(), Fn: dofn, Inbound: in, Out: out}
+					n.PID = transform.GetUniqueName()
+
+					input := unmarshalKeyedValues(transform.GetInputs())
+					for i := 1; i < len(input); i++ {
+						// TODO(herohde) 8/8/2018: handle different windows, view_fn and window_mapping_fn.
+						// For now, assume we don't need any information in the pardo payload.
+
+						ec, wc, err := b.makeCoderForPCollection(input[i])
+						if err != nil {
+							return nil, err
+						}
+
+						sid := StreamID{
+							Port:         Port{URL: b.desc.GetStateApiServiceDescriptor().GetUrl()},
+							PtransformID: id.to,
+						}
+						sideInputID := fmt.Sprintf("i%v", i) // SideInputID (= local id, "iN")
+						side := NewSideInputAdapter(sid, sideInputID, coder.NewW(ec, wc))
+						n.Side = append(n.Side, side)
 					}
-
-					sid := StreamID{
-						Port:         Port{URL: b.desc.GetStateApiServiceDescriptor().GetUrl()},
-						PtransformID: id.to,
+					if urn == urnProcessSizedElementsAndRestrictions {
+						u = &ProcessSizedElementsAndRestrictions{PDo: n}
+					} else {
+						u = n
 					}
```

Comment:
Consider moving the common case to always be used, and then corrected afterwards.
```suggestion
					u = n
					if urn == urnProcessSizedElementsAndRestrictions {
						u = &ProcessSizedElementsAndRestrictions{PDo: n}
					}
```

This idiom avoids the else block, and indicates that if there were a ?: operator in Go, that's what would happen. The compiler will optimize this appropriately.

Reply:
Done.